include/highfive/bits/H5Attribute_misc.hpp

```diff
@@ -67,10 +63,25 @@ inline void Attribute::read(T& array) const {
            << " into arrays of dimensions " << buffer_info.n_dimensions;
         throw DataSpaceException(ss.str());
     }
-    details::data_converter<T> converter(mem_space);
-    read(converter.transform_read(array), buffer_info.data_type);
+    auto dims = mem_space.getDimensions();
+    auto r = details::data_converter::get_reader<T>(dims);
+    read(r.get_pointer(), buffer_info.data_type);
     // re-arrange results
-    converter.process_result(array);
+    auto array = r.get_value();
+    auto t = create_datatype<typename details::inspector<T>::base_type>();
+    auto c = t.getClass();
+    if (c == DataTypeClass::VarLen) {
+        (void) H5Dvlen_reclaim(t.getId(), mem_space.getId(), H5P_DEFAULT, r.get_pointer());
+    }
+    return array;
+}
+
+template <typename T>
+inline void Attribute::read(T& array) const {
+    const DataSpace& mem_space = getMemSpace();
+    auto dims = mem_space.getDimensions();
+    details::inspector<T>::prepare(array, dims);
+    array = read<T>();
```

Comment:
These lines look counter intuitive to me. Therefore, a few questions:
* If a user allocates an `array` of the correct shape and passes this to `Attribute::read(T& array)`. Will these lines cause a move? If yes, what happens if `T` only support copying?
* Can we restructure this in such a way that `T Attribute::read()` becomes syntactic sugar for `void Attribute::read(T&)`. Currently it's the other way around. In the simplest case it would be
```
T Attribute::read() const {
    T array;

    // Since the sizes aren't as needed, this
    // will trigger re-allocation of `array`.
    read(array);

    return array;
}

Reply:
done :)